### PR TITLE
Introduce Thrust benchmarks

### PR DESCRIPTION
**Pull request by senior-zero**
## Description

closes https://github.com/NVIDIA/cccl/issues/473

This PR:
- Ports existing Thrust benchmarks into CCCL repository
- Extends Thrust benchmarks to work with any given device system (CUDA, CPP, OMP, etc.)
- Adds optional tests of benchmark data generators (can be enabled with `CUB_ENABLE_NVBENCH_HELPER_TESTS`)
- Refactors benchmark data generators API 
- Makes benchmark data generators heterogeneous
- Abstracts CUB tuning infrastructure from CUB and associates it with CCCL instead

To run Thrust benchmarks, it's sufficient to:
```bash
$: cmake -DCCCL_ENABLE_THRUST=YES \ 
      -DCCCL_ENABLE_CUB=YES \
      -DCCCL_ENABLE_LIBCUDACXX=NO \
      -DCMAKE_BUILD_TYPE=Release \
      -DCMAKE_CUDA_ARCHITECTURES=89 \
      -DCCCL_ENABLE_BENCHMARKS=YES .. 
$: ../benchmarks/scripts/run.py -R 'thrust.*'
&&&& RUNNING bench
 ctk:  12.2.140
cccl:  monorepo-309-g9ce36e3b7
&&&& PERF thrust_bench_adjacent_difference_basic_base_T_ct__I8___Elements_pow2__16 1.0048000149254221e-05 -sec
&&&& PERF thrust_bench_adjacent_difference_basic_base_T_ct__I8___Elements_pow2__20 0.00013096000475343317 -sec
&&&& PERF thrust_bench_adjacent_difference_basic_base_T_ct__I8___Elements_pow2__24 0.00017212801321875304 -sec
&&&& PERF thrust_bench_adjacent_difference_basic_base_T_ct__I8___Elements_pow2__28 0.0007995199994184077 -sec
&&&& PERF thrust_bench_adjacent_difference_basic_base_T_ct__I16___Elements_pow2__16 1.0239999937766697e-05 -sec
&&&& PERF thrust_bench_adjacent_difference_basic_base_T_ct__I16___Elements_pow2__20 0.0001341439929092303 -sec
&&&& PERF thrust_bench_adjacent_difference_basic_base_T_ct__I16___Elements_pow2__24 0.00021606399968732148 -sec
&&&& PERF thrust_bench_adjacent_difference_basic_base_T_ct__I16___Elements_pow2__28 0.0014510079054161906 -sec
```

Thrust multiconfig setup is used to produce benchmarks for different device systems. 

## Checklist

- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Review by miscco — APPROVED** [review]

**Review by allisonvacanti — APPROVED** [review]

**Review by wmaxey — APPROVED** [review]

File: cub/benchmarks/nvbench_helper/test/gen_entropy.cu [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+                                     d_in,
+                                     d_unique_out,
+                                     d_counts_out,
+                                     d_num_runs_out,
+                                     n);
+
+  thrust::device_vector<std::uint8_t> temp_storage(temp_storage_bytes);
+  d_temp_storage = thrust::raw_pointer_cast(temp_storage.data());
+
+  cub::DeviceRunLengthEncode::Encode(d_temp_storage,
```
Comment (elstehle) [inline cub/benchmarks/nvbench_helper/test/gen_entropy.cu:92]: question: Wouldn't we need to pre-sort `in` before running `cub::DeviceRunLengthEncode::Encode` here?

File: cub/benchmarks/nvbench_helper/test/gen_entropy.cu [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+
+  T *d_in             = thrust::raw_pointer_cast(in.data());
+  T *d_unique_out     = thrust::raw_pointer_cast(unique.data());
+  int *d_counts_out   = thrust::raw_pointer_cast(counts.data());
+  int *d_num_runs_out = thrust::raw_pointer_cast(num_runs.data());
+
+  cub::DeviceRunLengthEncode::Encode(d_temp_storage,
+                                     temp_storage_bytes,
+                                     d_in,
+                                     d_unique_out,
```
Comment (elstehle) [inline cub/benchmarks/nvbench_helper/test/gen_entropy.cu:84]: nit:
```suggestion
                                     thrust::make_discard_iterator(),
```

File: cub/benchmarks/nvbench_helper/test/gen_entropy.cu [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+                                     d_num_runs_out,
+                                     n);
+
+  thrust::device_vector<std::uint8_t> temp_storage(temp_storage_bytes);
+  d_temp_storage = thrust::raw_pointer_cast(temp_storage.data());
+
+  cub::DeviceRunLengthEncode::Encode(d_temp_storage,
+                                     temp_storage_bytes,
+                                     d_in,
+                                     d_unique_out,
```
Comment (elstehle) [inline cub/benchmarks/nvbench_helper/test/gen_entropy.cu:95]: nit:
```suggestion
                                     thrust::make_discard_iterator(),
```

File: cub/benchmarks/nvbench_helper/test/gen_entropy.cu [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+  thrust::device_vector<int> counts(n);
+  thrust::device_vector<int> num_runs(1);
+  thrust::sort(in.begin(), in.end(), less_t{});
+
+  // RLE
+  void *d_temp_storage           = nullptr;
+  std::size_t temp_storage_bytes = 0;
+
+  T *d_in             = thrust::raw_pointer_cast(in.data());
+  T *d_unique_out     = thrust::raw_pointer_cast(unique.data());
```
Comment (elstehle) [inline cub/benchmarks/nvbench_helper/test/gen_entropy.cu:77]: nit: we could get rid of this if we use a `discard_iterator` here
```suggestion
```

File: cub/benchmarks/nvbench_helper/test/gen_entropy.cu [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+  return sizeof(T) * 8 * entropy;
+}
+
+template <class T>
+double compute_actual_entropy(thrust::device_vector<T> in)
+{
+  const int n = static_cast<int>(in.size());
+
+  #if THRUST_DEVICE_SYSTEM == THRUST_DEVICE_SYSTEM_CUDA
+  thrust::device_vector<T> unique(n);
```
Comment (elstehle) [inline cub/benchmarks/nvbench_helper/test/gen_entropy.cu:67]: nit: we could get rid of this if we use a `discard_iterator` here
```suggestion
```

**Review by elstehle — APPROVED** [review]
Thanks for the great work! That's a awesome step up for our benchmarking infrastructure! 

Just a few, mostly minor, comments/suggestions.

File: cub/benchmarks/nvbench_helper/test/gen_range.cu [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+TEMPLATE_LIST_TEST_CASE("Generators produce data within specified range", "[gen]", types)
+{
+  const auto min = static_cast<TestType>(GENERATE_COPY(take(3, random(-124, 0))));
+  const auto max = static_cast<TestType>(GENERATE_COPY(take(3, random(0, 124))));
+
+  const thrust::device_vector<TestType> data = generate(1 << 16, bit_entropy::_1_000, min, max);
+
+  const TestType min_element = *thrust::min_element(data.begin(), data.end());
+  const TestType max_element = *thrust::min_element(data.begin(), data.end());
```
[the remark this answers is not in the document]
Comment (senior-zero) [reply cub/benchmarks/nvbench_helper/test/gen_range.cu:47]: Good catch!

File: cub/benchmarks/nvbench_helper/test/gen_entropy.cu (hunk 3 above)
Comment (senior-zero) [reply cub/benchmarks/nvbench_helper/test/gen_entropy.cu:95]: https://github.com/NVIDIA/cccl/issues/556

**Pull request merged**